### PR TITLE
cleanup(flake): add retry loop to mitigate flakiness

### LINKED ISSUE #9728: [Flake] bigquery_read_integration_test CreateReadSession() internal error
```
2022-08-24T18:38:54.587112000Z [DEBUG] <139951507649856> CreateReadSession() << google.cloud.bigquery.storage.v1.CreateReadSessionRequest {
  parent: "projects/cloud-cpp-testing-resources"
  read_session {
    data_format: AVRO
    table: "projects/bigquery-public-data/datasets/usa_names/tables/usa_1910_current"
    read_options {
      row_restriction: "state = \"WA\""
    }
  }
} (./google/cloud/internal/log_wrapper.h:91)
2022-08-24T18:39:54.691494000Z [DEBUG] <139951507649856> CreateReadSession() >> status=INTERNAL: request failed: internal error (./google/cloud/internal/log_wrapper.h:94)
```

* GCB: https://console.cloud.google.com/cloud-build/builds;region=us-east1/e8d5ce8b-0186-49c9-aed1-5d50fd05fc23;tab=detail?project=cloud-cpp-testing-resources
* Raw: https://storage.googleapis.com/cloud-cpp-community-publiclogs/logs/google-cloud-cpp/main/ea9c8db638027b5fbdb238fdb1c2344a6de763b4/fedora-36-ubsan/log-e8d5ce8b-0186-49c9-aed1-5d50fd05fc23.txt

### PR DESCRIPTION
**Pull request by scotthart**
fixes #9728

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9752)

### CODE REVIEW

File: google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc [hunk 1]
```diff
@@ -80,6 +81,19 @@ std::int64_t CountRowsFromStream(
   return num_rows;
 }
 
+template <typename Functor>
+google::cloud::internal::invoke_result_t<Functor> StatusRetryLoop(
+    Functor&& operation,
+    std::vector<int> const& delay_seconds = {10, 30, 60, 60}) {
```
Comment (devbww) [inline google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc:87]: Nit: I'd suggest a vector of durations.  That is, convert from `int` as early as possible.

File: google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc [hunk 2]
```diff
@@ -80,6 +81,19 @@ std::int64_t CountRowsFromStream(
...
+  google::cloud::internal::invoke_result_t<Functor> status;
+  for (auto delay : delay_seconds) {
+    status = operation();
+    if (status.ok()) break;
```
Comment (devbww) [inline google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc:91]: Q. Is there some distinction between retryable and non-retryable errors that we should make?  Can we use retry/backoff policies instead.

File: google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc [hunk 3]
```diff
@@ -177,7 +191,9 @@ TEST_F(BigQueryReadIntegrationTest, ReadRowsSuccess) {
       "usa_1910_current");
   read_session.mutable_read_options()->set_row_restriction("state = \"WA\"");
   *session_request.mutable_read_session() = read_session;
-  auto session_response = client.CreateReadSession(session_request);
+  auto session_response = StatusRetryLoop(
+      [&] { return client.CreateReadSession(session_request); });
```
Comment (devbww) [inline google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc:195]: Q. How come this only applies to this particular `CreateReadSession()` call?

File: google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc [hunk 4]
```diff
@@ -177,7 +191,9 @@ TEST_F(BigQueryReadIntegrationTest, ReadRowsSuccess) {
...
+
```
Comment (devbww) [inline google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc:196]: Nit: Remove new blank line?

File: google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc (hunk 2 above)
Comment (devbww) [inline google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc:91]: Q. If we know that `status.ok()` is a valid expression, should we not abstract away `Status` from the functor signature?

Comment (coryan) [reply google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc:91]: ➕1  You could use the existing retry loop with a custom idempotency policy:

```cc
class TestIdempotencyPolicy : public BigQueryReadConnectionIdempotencyPolicy {
 public:
  virtual ~TestIdempotencyPolicy() override = default;

  /// Create a new copy of this object.
  std::unique_ptr<BigQueryReadConnectionIdempotencyPolicy> clone() const override {
    return absl::make_unique<TestIdempotencyPolicy>();
  }

  google::cloud::Idempotency CreateReadSession(
      google::cloud::bigquery::storage::v1::CreateReadSessionRequest const&) override {
    return google::cloud::kIdempotent;
  }

  google::cloud::Idempotency SplitReadStream(
      google::cloud::bigquery::storage::v1::SplitReadStreamRequest const&) override {
    return google::cloud::kIdempotent;
  }
};
```


And initialize the `*Connection` with that policy.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9752)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @coryan and @devbww)

___
*[`google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc` line 87 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9752#-NAkxXRkiijDf1v5NQKv-r1-87:-NAov5dG61T6YtJASppU:b-4ctsav) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2433cab7f4a7a73af33f65b895a96757a6ab1d12/google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc#L87)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: I'd suggest a vector of durations.  That is, convert from `int` as early as possible.
</blockquote></details>

Removed this function.
___
*[`google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc` line 91 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9752#-NAkxXRkiijDf1v5NQKv-r1-91:-NAov7XO-UjEiG6uRwnX:b-q0myzi) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2433cab7f4a7a73af33f65b895a96757a6ab1d12/google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc#L91)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

➕1  You could use the existing retry loop with a custom idempotency policy:

```cc
class TestIdempotencyPolicy : public BigQueryReadConnectionIdempotencyPolicy {
 public:
  virtual ~TestIdempotencyPolicy() override = default;

  /// Create a new copy of this object.
  std::unique_ptr<BigQueryReadConnectionIdempotencyPolicy> clone() const override {
    return absl::make_unique<TestIdempotencyPolicy>();
  }

  google::cloud::Idempotency CreateReadSession(
      google::cloud::bigquery::storage::v1::CreateReadSessionRequest const&) override {
    return google::cloud::kIdempotent;
  }

  google::cloud::Idempotency SplitReadStream(
      google::cloud::bigquery::storage::v1::SplitReadStreamRequest const&) override {
    return google::cloud::kIdempotent;
  }
};
```


And initialize the `*Connection` with that policy.

</blockquote></details>

Using Idempotency, Retry, and Backoff Policy options in lieu of adding the new function.
___
*[`google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc` line 195 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9752#-NAkxXRkiijDf1v5NQKv-r1-195:-NAovb3v7bqJXo4LP9PT:bl2cdzh) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2433cab7f4a7a73af33f65b895a96757a6ab1d12/google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc#L195)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Q. How come this only applies to this particular `CreateReadSession()` call?
</blockquote></details>

Not sure why this sometimes succeeds and sometimes fails. Hopefully retrying will reduce the flakiness.
___
*[`google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc` line 196 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9752#-NAkxXRkiijDf1v5NQKv-r1-196:-NAov_g29ScMr5H5ToLZ:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2433cab7f4a7a73af33f65b895a96757a6ab1d12/google/cloud/bigquery/integration_tests/bigquery_read_integration_test.cc#L196)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Nit: Remove new blank line?
</blockquote></details>

Done.

**Review by devbww — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9752)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan)

**Review by coryan — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9752)* status: all files reviewed, 1 unresolved discussion (waiting on @devbww)

**Pull request merged**